LLM Semantic Conventions: Initial PR (#825)
open-telemetry/semantic-conventions: docs/gen-ai/llm-spans.md

<!--- Hugo front matter used to generate the website version of this page:
linkTitle: LLM Calls
--->

# Semantic Conventions for LLM requests

**Status**: [Experimental][DocumentStatus]

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Configuration](#configuration)
- [LLM Request attributes](#llm-request-attributes)
- [Events](#events)

<!-- tocstop -->

A request to an LLM is modeled as a span in a trace.

> The **span name** SHOULD be set to a low cardinality value representing the request made to an LLM.
> It MAY be a name of the API endpoint for the LLM being called.

Reviewer: We don't usually use endpoints in span names, I'm suggesting to use API name

```suggestion
The **span name** SHOULD be set to a low cardinality value describing an operation made to an LLM. 
For example, the API name such as [Create chat completion](https://platform.openai.com/docs/api-reference/chat/create) 
```

I think we should be more specific here eventually , but I'm fine keeping it vague for now

Reviewer 2: Suggestion is to use the request type and model identifier such as "ChatCompletion Gpt4".

## Configuration

> Instrumentations for LLMs MUST offer the ability to turn off capture of prompts and completions. This is for three primary reasons:

Reviewer: I think we should allow instrumentations that don't want to capture prompts and completions do their job.

```suggestion
Instrumentations MAY capture prompts and completions. Instrumentations that support it, MUST offer the ability to turn off capture of prompts and completions. This is for three primary reasons:
```

Reviewer 3: Prefer using feature flags to turn this on and off, as with some LLMs/VectorDBs, different personas may have different requirements for this: Application Developer / Data Scientist may want to see prompt and response for debug purpose while end users sometime may want to hide those values for security reason, make sense? Thanks!

Reviewer 2: > Prefer using feature flags to turn this on and off, as with some LLMs/VectorDBs, different personas may have different requirements for this: Application Developer / Data Scientist may want to see prompt and response for debug purpose while end users sometime may want to hide those values for security reason, make sense? Thanks!

The implementation will be up to the instrumentation library, but agree that this should be off by default and then turned on by some opt-in mechanism.

1. Data privacy concerns. End users of LLM applications may input sensitive information or personally identifiable information (PII) that they do not wish to be sent to a telemetry backend.
2. Data size concerns. Although there is no specified limit to sizes, there are practical limitations in programming languages and telemety systems. Some LLMs allow for extremely large context windows that end users may take full advantage of.
3. Performance concerns. Sending large amounts of data to a telemetry backend may cause performance issues for the application.

> By default, these configurations SHOULD NOT capture prompts and completions.

Reviewer: I think there is a contradiction between this sentence, the requirement level on the corresponding attributes, and the previous paragraph. 

If we want prompts/completions to be off by default, the attributes should have opt-in level. 
Then we previous paragraph should be rephrased to something like

"Instrumentations MAY capture prompts and completions. They contain sensitive and verbose content and therefore are only captured if enabled explicitly - see [Attribute Requirement Levels - Opt-in](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/#opt-in)"  for the details.

I *think* we can entertain the idea of more relaxed opt-in. E.g. if instrumentation can reliably detect that code is running inside IDE or in development environment, etc. Let's see.

<!-- semconv gen_ai.llm.request -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| [`gen_ai.llm.request.max_tokens`](../attributes-registry/gen-ai.md) | int | The maximum number of tokens the LLM generates for a request. | `100` | Recommended |
| [`gen_ai.llm.request.model`](../attributes-registry/gen-ai.md) | string | The name of the LLM a request is being made to. [1] | `gpt-4` | Required |
| [`gen_ai.llm.request.temperature`](../attributes-registry/gen-ai.md) | double | The temperature setting for the LLM request. | `0.0` | Recommended |
| [`gen_ai.llm.request.top_p`](../attributes-registry/gen-ai.md) | double | The top_p sampling setting for the LLM request. | `1.0` | Recommended |
| [`gen_ai.llm.response.finish_reason`](../attributes-registry/gen-ai.md) | string[] | Array of reasons the model stopped generating tokens, corresponding to each generation received. | `[['stop']]` | Recommended |
| [`gen_ai.llm.response.id`](../attributes-registry/gen-ai.md) | string | The unique identifier for the completion. | `chatcmpl-123` | Recommended |

> | [`gen_ai.llm.response.model`](../attributes-registry/gen-ai.md) | string | The name of the LLM a response is being made to. [2] | `gpt-4-0613` | Required |

Reviewer: the could be no response if the was an error, so nothing on the response namespace can be required.

I suggest something like 
```suggestion
| [`gen_ai.llm.response.model`](../attributes-registry/gen-ai.md) | string | The name of the LLM a response is being made to. [2] | `gpt-4-0613` | Conditionally required: if response was received |
```

> | [`gen_ai.llm.system`](../attributes-registry/gen-ai.md) | string | The name of the LLM foundation model vendor, if applicable. [3] | `openai` | Recommended |

Reviewer: We tend to make `*.system` a  required attribute. It's necessary to 
- differentiate telemetry about LLM from all other spans  - if `llm.system` is present, I know it is. Otherwise things get complicated
- as a discriminator allowing to search for system-specific attributes such as `llm.openai.*` ones. 

```suggestion
| [`gen_ai.llm.system`](../attributes-registry/gen-ai.md) | string | The name of the LLM foundation model vendor, if applicable. [3] | `openai` | Required|
```

So I wonder what are the cases where we don't know the client-library view of the system it talks to?
I remember the concern around it - https://github.com/open-telemetry/semantic-conventions/issues/813, let's follow up there.

**[1]:** The name of the LLM a request is being made to. If the LLM is supplied by a vendor, then the value must be the exact name of the model requested. If the LLM is a fine-tuned custom model, the value should have a more specific name than the base model that's been fine-tuned.

**[2]:** The name of the LLM a response is being made to. If the LLM is supplied by a vendor, then the value must be the exact name of the model actually used. If the LLM is a fine-tuned custom model, the value should have a more specific name than the base model that's been fine-tuned.

**[3]:** The name of the LLM foundation model vendor, if applicable. If not using a vendor-supplied model, this field is left blank.
<!-- endsemconv -->

## Events

In the lifetime of an LLM span, an event for prompts sent and completions received MAY be created, depending on the configuration of the instrumentation.

<!-- semconv gen_ai.llm.content.prompt -->
The event name MUST be `gen_ai.llm.content.prompt`.

| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|

> | [`gen_ai.llm.prompt`](../attributes-registry/gen-ai.md) | string | The full prompt sent to an LLM, as a stringified JSON in OpenAI's format. [1] | `[{'role': 'user', 'content': 'What is the capital of France?'}]` | Recommended |

Reviewer: I have some ideas on how we can document events better and maybe avoid the span events vs log events trap.
I'll send a PR against your branch so it's easier to discuss

Reviewer: I was referring to this requirement level in one of the previous comments, I think this should be opt-in. As an alternative, we should consider the whole event as opt-in. We can try adding the feature to the yaml to support it, but that's not possible today

<!--- Hugo front matter used to generate the website version of this page:
linkTitle: LLM Calls
--->

# Semantic Conventions for LLM requests

**Status**: [Experimental][DocumentStatus]

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Configuration](#configuration)
- [LLM Request attributes](#llm-request-attributes)
- [Events](#events)

<!-- tocstop -->

A request to an LLM is modeled as a span in a trace.

> 

Reviewer 2: ```suggestion

**Span kind:** MUST always be `CLIENT`.

```

The **span name** SHOULD be set to a low cardinality value describing an operation made to an LLM.

> For example, the API name such as [Create chat completion](https://platform.openai.com/docs/api-reference/chat/create)

Reviewer 2: ```suggestion
For example, the API name such as [Create chat completion](https://platform.openai.com/docs/api-reference/chat/create) could be represented as `ChatCompletions gpt-4` to represent the API and the LLM.
```
I think completing this sentence with an example would help. In a future PR, we could include additional examples.

<!--- Hugo front matter used to generate the website version of this page:
linkTitle: LLM requests
--->

# Semantic Conventions for LLM requests

**Status**: [Experimental][DocumentStatus]

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Configuration](#configuration)
- [LLM Request attributes](#llm-request-attributes)
- [Events](#events)

<!-- tocstop -->

A request to an LLM is modeled as a span in a trace.

**Span kind:** MUST always be `CLIENT`.

The **span name** SHOULD be set to a low cardinality value describing an operation made to an LLM.
For example, the API name such as [Create chat completion](https://platform.openai.com/docs/api-reference/chat/create) could be represented as `ChatCompletions gpt-4` to include the API and the LLM.

## Configuration

> Instrumentations for LLMs MAY capture prompts and completions.

Author: I don't think it must be disabled. I found that most folks are interested in seeing these and expecting that. So I think there should be an option to disable, but ON by default.

Instrumentations that support it, MUST offer the ability to turn off capture of prompts and completions. This is for three primary reasons:

1. Data privacy concerns. End users of LLM applications may input sensitive information or personally identifiable information (PII) that they do not wish to be sent to a telemetry backend.

> 2. Data size concerns. Although there is no specified limit to sizes, there are practical limitations in programming languages and telemety systems. Some LLMs allow for extremely large context windows that end users may take full advantage of.

Reviewer: ```suggestion
2. Data size concerns. Although there is no specified limit to sizes, there are practical limitations in programming languages and telemetry systems. Some LLMs allow for extremely large context windows that end users may take full advantage of.
```

| [`gen_ai.response.id`](../attributes-registry/llm.md) | string | The unique identifier for the completion. | `chatcmpl-123` | `Recommended` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| [`gen_ai.response.model`](../attributes-registry/llm.md) | string | The name of the LLM a response was generated from. [3] | `gpt-4-0613` | `Recommended` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| [`gen_ai.usage.completion_tokens`](../attributes-registry/llm.md) | int | The number of tokens used in the LLM response (completion). | `180` | `Recommended` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| [`gen_ai.usage.prompt_tokens`](../attributes-registry/llm.md) | int | The number of tokens used in the LLM prompt. | `100` | `Recommended` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

**[1]:** The name of the LLM a request is being made to. If the LLM is supplied by a vendor, then the value must be the exact name of the model requested. If the LLM is a fine-tuned custom model, the value should have a more specific name than the base model that's been fine-tuned.

> **[2]:** If not using a vendor-supplied model, provide a custom friendly name, such as a name of the company or project. If the instrumetnation reports any attributes specific to a custom model, the value provided in the `gen_ai.system` SHOULD match the custom attribute namespace segment. For example, if `gen_ai.system` is set to `the_best_llm`, custom attributes should be added in the `gen_ai.the_best_llm.*` namespace. If none of above options apply, the instrumentation should set `_OTHER`.

Reviewer: ```suggestion
**[2]:** If not using a vendor-supplied model, provide a custom friendly name, such as a name of the company or project. If the instrumentation reports any attributes specific to a custom model, the value provided in the `gen_ai.system` SHOULD match the custom attribute namespace segment. For example, if `gen_ai.system` is set to `the_best_llm`, custom attributes should be added in the `gen_ai.the_best_llm.*` namespace. If none of above options apply, the instrumentation should set `_OTHER`.
```
